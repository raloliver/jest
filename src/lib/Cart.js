import find from 'lodash/find';
import remove from 'lodash/remove';

export default class Cart {
  items = [];

  add(item) {
    const itemDuplicated = {product: item.product};

    if (find(this.items, itemDuplicated)) {
      remove(this.items, itemDuplicated);
    }

    this.items.push(item);
  }

  remove(product) {
    remove(this.items, {product});
  }

  summary() {
    const total = this.getTotal();
    const items = this.items;

    return {
      total,
      items,
    };
  }

  checkout() {
    const {total, items} = this.summary();

    this.items = [];

    return {
      total,
      items,
    };
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }
}
