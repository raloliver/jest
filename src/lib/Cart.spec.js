import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let product = {
    title: 'Nintendo Switch',
    price: 263,
  };

  let product1 = {
    title: 'Ã“culos Rift',
    price: 300,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  it('should return 0 when getTotal() init', () => {
    const cart = new Cart();
    expect(cart.getTotal()).toEqual(0);
  });

  it('should multiply quantity and price and return the total amount', () => {
    const item = {
      product,
      quantity: 2,
    };

    cart.add(item);
    expect(cart.getTotal()).toEqual(526);
  });

  it('should ensure only one type of product on the cart', () => {
    cart.add({
      product,
      quantity: 2,
    });
    cart.add({
      product,
      quantity: 1,
    });
    expect(cart.getTotal()).toEqual(263);
  });

  it('should update total when a product gets included and then removed', () => {
    cart.add({
      product,
      quantity: 2,
    });
    cart.add({
      product: product1,
      quantity: 2,
    });
    cart.remove(product);
    expect(cart.getTotal()).toEqual(600);
  });
});
