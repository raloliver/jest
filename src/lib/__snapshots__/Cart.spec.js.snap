// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should return an object with the total and the items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 263,
        "title": "Nintendo Switch",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 300,
        "title": "Óculos Rift",
      },
      "quantity": 5,
    },
  ],
  "total": 2026,
}
`;

exports[`Cart checkout should return an object with the total and the items when summary() 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 263,
        "title": "Nintendo Switch",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 300,
        "title": "Óculos Rift",
      },
      "quantity": 1,
    },
  ],
  "total": 826,
}
`;
